extensions/amp-next-page/0.2/visibility-observer.js

```diff
@@ -0,0 +1,188 @@
+      return;
+    }
+    this.bottomSentinelPosition_ = position.relativePos;
+    this.updateRelativePos_(callback);
+  }
+
+  /**
+   * Calculates the position of the element relative to the viewport
+   * based on the positions of the injected bottom and top sentinel elements
+   * @param {function(!ViewportRelativePos)} callback
+   * @private
+   */
+  updateRelativePos_(callback) {
+    const {topSentinelPosition_: top, bottomSentinelPosition_: bottom} = this;
+    const {INSIDE, TOP, BOTTOM} = RelativePositions;
+
+    devAssert(
+      top || bottom,
+      'scroll triggered without a top or bottom sentinel position'
+    );
+
+    if (top === INSIDE && bottom === INSIDE) {
+      // Both the top and bottom sentinel elements are within the
+      // viewport bounds meaning that the document is short enough
+      // to be contained inside the viewport
+      this.relativePos_ = ViewportRelativePos.INSIDE_VIEWPORT;
+    } else if ((!top || top === TOP) && (!bottom || bottom === BOTTOM)) {
+      // The head of the document is above the viewport and the
+      // foot of the document is below it, meaning that the viewport
+      // is looking at a section of the document
+      this.relativePos_ = ViewportRelativePos.CONTAINS_VIEWPORT;
+    } else if (
+      ((!top || top === TOP) && bottom === TOP) ||
+      (top === BOTTOM && (!bottom || bottom === BOTTOM))
+    ) {
+      // Both the top and the bottom of the document are either
+      // above or below the document meaning that the viewport hasn't
+      // reached the document yet or has passed it
+      this.relativePos_ = ViewportRelativePos.OUTSIDE_VIEWPORT;
+    } else {
+      // The remaining case is the case where the document is halfway
+      // through being scrolling into/out of the viewport in which case
+      // we don't need to update the visibility
+      this.relativePos_ = ViewportRelativePos.LEAVING_VIEWPORT;
+      return;
+    }
+
+    callback(this.relativePos_);
+  }
+}
+
+export default class VisibilityObserver {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!PositionObserver=} injectedPositionObserver
+   */
+  constructor(ampdoc, injectedPositionObserver) {
+    /** @private @const {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+    /** @private @const {Array<!VisibilityObserverEntry>} */
+    this.entries_ = [];
+    /** @private @const {?PositionObserver} */
+    this.injectedPositionObserver_ = injectedPositionObserver || null;
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {!Element} parent
+   * @param {function(!ViewportRelativePos)} callback
+   */
+  observe(element, parent, callback) {
+    const entry = new VisibilityObserverEntry(this);
+    this.entries_.push(entry);
+    entry.observe(element, parent, callback);
+  }
+
+  /**
+   * @return {!PositionObserver}
+   */
+  getPositionObserver() {
+    // For testing
+    if (this.injectedPositionObserver_) {
```

Comment:
I think the more common pattern in AMP would be to do something like: https://github.com/ampproject/amphtml/blob/168b94bb0e4e2f295ba5c1017a239f51cd6fa7a1/test/unit/test-xhr.js#L150-L154

Would suggest looking into that since it is a more common pattern.